cannon/mipsevm/state_mt.go

```diff
@@ -0,0 +1,293 @@
+	FutexAddr        uint32     `json:"futexAddr"`
+	FutexVal         uint32     `json:"futexVal"`
+	FutexTimeoutStep uint64     `json:"futexTimeoutStep"`
+	Cpu              CpuScalars `json:"cpu"`
+	Registers        [32]uint32 `json:"registers"`
+}
+
+func (t *ThreadState) serializeThread() []byte {
+	// TODO
+	out := make([]byte, 0, SERIALIZED_THREAD_SIZE)
+
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.PC)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.NextPC)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.LO)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.HI)
+	//
+	//for _, r := range t.Registers {
+	//	out = binary.BigEndian.AppendUint32(out, r)
+	//}
+
+	return out
+}
+
+func computeThreadRoot(prevStackRoot common.Hash, threadToPush *ThreadState) common.Hash {
+	hashedThread := crypto.Keccak256Hash(threadToPush.serializeThread())
+
+	var hashData []byte
+	hashData = append(hashData, prevStackRoot[:]...)
+	hashData = append(hashData, hashedThread[:]...)
+
+	return crypto.Keccak256Hash(hashData)
+}
+
+// MT_STATE_WITNESS_SIZE is the size of the state witness encoding in bytes.
+const MT_STATE_WITNESS_SIZE = 163
+const (
+	MEMROOT_MT_WITNESS_OFFSET                    = 0
+	PREIMAGE_KEY_MT_WITNESS_OFFSET               = MEMROOT_MT_WITNESS_OFFSET + 32
+	PREIMAGE_OFFSET_MT_WITNESS_OFFSET            = PREIMAGE_KEY_MT_WITNESS_OFFSET + 32
+	HEAP_MT_WITNESS_OFFSET                       = PREIMAGE_OFFSET_MT_WITNESS_OFFSET + 4
+	EXITCODE_MT_WITNESS_OFFSET                   = HEAP_MT_WITNESS_OFFSET + 4
+	EXITED_MT_WITNESS_OFFSET                     = EXITCODE_MT_WITNESS_OFFSET + 1
+	STEP_MT_WITNESS_OFFSET                       = EXITED_MT_WITNESS_OFFSET + 1
+	STEPS_SINCE_CONTEXT_SWITCH_MT_WITNESS_OFFSET = STEP_MT_WITNESS_OFFSET + 8
+	WAKEUP_MT_WITNESS_OFFSET                     = STEPS_SINCE_CONTEXT_SWITCH_MT_WITNESS_OFFSET + 8
+	TRAVERSE_RIGHT_MT_WITNESS_OFFSET             = WAKEUP_MT_WITNESS_OFFSET + 4
+	LEFT_THREADS_ROOT_MT_WITNESS_OFFSET          = TRAVERSE_RIGHT_MT_WITNESS_OFFSET + 1
+	RIGHT_THREADS_ROOT_MT_WITNESS_OFFSET         = LEFT_THREADS_ROOT_MT_WITNESS_OFFSET + 32
+	THREAD_ID_MT_WITNESS_OFFSET                  = RIGHT_THREADS_ROOT_MT_WITNESS_OFFSET + 32
+)
+
+type MTState struct {
+	Memory *Memory `json:"memory"`
+
+	PreimageKey    common.Hash `json:"preimageKey"`
+	PreimageOffset uint32      `json:"preimageOffset"` // note that the offset includes the 8-byte length prefix
+
+	Heap uint32 `json:"heap"` // to handle mmap growth
+
+	ExitCode uint8 `json:"exit"`
+	Exited   bool  `json:"exited"`
+
+	Step                        uint64 `json:"step"`
+	StepsSinceLastContextSwitch uint64 `json:"stepsSinceLastContextSwitch"`
+	Wakeup                      uint32 `json:"wakeup"`
+
+	TraverseRight         bool          `json:"traverseRight"`
+	LeftThreadStack       []ThreadState `json:"leftThreadStack"`
+	RightThreadStack      []ThreadState `json:"rightThreadStack"`
+	LeftThreadStackRoots  []common.Hash `json:"leftThreadStackRoots"`
```

Comment:
this is a nit since it's implementation specific comment. I wonder if it's easier to compute the roots lazily when creating state roots and witness? This would be less error-prone only computing roots when they're needed.

Reply:
I was initially thinking that we don't want to be rehashing everything repeatedly when we have an unbounded number of threads, but I suppose we only need to do this periodically whenever we generate snapshots or single step witnesses?  In which case, yeah, probably makes sense to simplify 👍 .

Reply:
Yup, snapshots occurs only on a tiny fraction of steps. Which is the only time we need the thread roots.